### DISCUSSION TITLE
OOM in count queries

### DISCUSSION THREAD

**Discussion by wey-gu** [vesoft-inc/nebula#4782] [Q&A]
I quoted DM session with @wdroste

> 
> I'm running a cluster with 5 nodes which are 4cpu 8gb ram and 250GB of nvme
> create space vid int64 w/ replication 3
> 5 nodes are c6id.xlarge and pushing data w/ t3.small
> 8 threads writing random data w/ indexes
> and trying to write 10m vertexes
> i'm kind of getting through the write but it seem slike i can break the cluster pretty easily
> i change storage to the following
> ```
> # Network IO threads number
> --num_io_threads=2
> # Worker threads number to handle request
> --num_worker_threads=4
> # Maximum subtasks to run admin jobs concurrently
> --max_concurrent_subtasks=1
> --rocksdb_rate_limit=20
> ```
> and the graphd dies w/ the lookup query
> 
> ```
> LOOKUP ON tagFoo YIELD id(vertex) | YIELD COUNT(*);
> ```
> 
> This was just a bit of a test, we do a lot of count queries for reports, so i was trying to see where the system would go. We do lots of traversals and well which i assume will go faster because there's a starting node..
> 
> the memory profile was great <1GB of ram but rocksdb killed storaged on compaction and i don't know if that's because of the memory? There was 7GB in buffer cache..

**Comment by wey-gu** [comment]
> and the graphd dies w/ the lookup query
>
> `LOOKUP ON tagFoo YIELD id(vertex) | YIELD COUNT(*);`

This is due to the count is not yet pushdown to the storage side, which will end up lookup to do a full index scan(as no filter condition here) and then count, this is not in hi-priority due to in full scan case we have other ways to get count, and in other cases, the scanned data will be filtered.

In this case(whole tag count), it’s recommended to use this async approach instead: SHOW STATS/SUBMIT JOB STATS

**Comment by wey-gu** [comment]
> the memory profile was great <1GB of ram but rocksdb killed storaged on compaction and i don't know if that's because of the memory? There was 7GB in buffer cache.

@wdroste regarding this, do you meant the storaged got killed during LOOKUP, too, or during a compaction only?

**Reply by wdroste** [reply]
I was doing the `LOOKUP` while loading the data to see where things might break, but it was definitely during a compaction that it core dumped.

**Comment by critical27** [comment]
You probably need to lower the configurations related to rocksdb, such as size of memtable, sst, thread num of background job.

**Reply by wey-gu** [reply]
@critical27 you meant those memtable/cache of rocksdb for all parts/shards actually consumed all RAM including the 7GiB buff/cache, right?

**Reply by wdroste** [reply]
I was starting down that path looking for the correct documentation around rocksdb was also just going to get apply a much bigger system. Can you please point me to the best location for configuration of the relative fields. RockDB's website is sparse on configuration or at least I can't find much.

**Reply by wey-gu** [reply]
> such as size of memtable, sst, thread num of background job

rocksdb related:

- memtable: `ColumnFamilyOptions::write_buffer_size` size for each memtable(default 64MiB), `AdvancedColumnFamilyOptions::max_write_buffer_number` max number of memtable, `DBOptions::db_write_buffer_size` all memtable size (default unlimited)
- sst: `target_file_size_base` size of sst file for level 1, `max_bytes_for_level_base` all level sst size
- num of background job: `max_background_jobs`

@wdroste also I put nebula-storaged configuration docs here and quote related chapter:

https://docs.nebula-graph.io/3.2.1/5.configurations-and-logs/1.configurations/4.storage-config/#rocksdb_options

<h2 id="rocksdb_options" style="box-sizing: inherit; font-weight: 300; letter-spacing: -0.01em; font-size: 1.5625em; line-height: 1.4; margin: 1.6em 0px 0.64em; --md-scroll-margin:3.6rem; --md-scroll-offset:0.2rem; scroll-margin-top: calc(var(--md-scroll-margin) - var(--md-scroll-offset)); color: rgb(0, 0, 0); font-family: Roboto, -apple-system, &quot;system-ui&quot;, Helvetica, Arial, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">RocksDB options<a class="headerlink" href="https://docs.nebula-graph.io/3.2.1/5.configurations-and-logs/1.configurations/4.storage-config/#rocksdb_options" title="Permanent link" style="box-sizing: inherit; -webkit-tap-highlight-color: transparent; color: var(--md-accent-fg-color); text-decoration: none; word-break: break-word; transition: color 0.25s ease 0s, opacity 125ms ease 0s; display: inline-block; opacity: 1; margin-left: 0.5rem;">¶</a></h2><div class="md-typeset__scrollwrap" style="box-sizing: inherit; margin: 1em -0.8rem; overflow-x: auto; touch-action: auto; color: rgb(0, 0, 0); font-family: Roboto, -apple-system, &quot;system-ui&quot;, Helvetica, Arial, sans-serif; font-size: 17.6px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="md-typeset__table" style="box-sizing: inherit; display: inline-block; margin-bottom: 0.5em; padding: 0px 0.8rem;">

Name | Predefined value | Description
-- | -- | --
rocksdb_db_options | {} | Specifies the RocksDB database options.
rocksdb_column_family_options | {"write_buffer_size":"67108864","max_write_buffer_number":"4","max_bytes_for_level_base":"268435456"} | Specifies the RocksDB column family options.
rocksdb_block_based_table_options | {"block_size":"8192"} | Specifies the RocksDB block based table options.

</div></div><p style="box-sizing: inherit; color: rgb(0, 0, 0); font-family: Roboto, -apple-system, &quot;system-ui&quot;, Helvetica, Arial, sans-serif; font-size: 17.6px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">The format of the RocksDB option is<span> </span><code style="box-sizing: inherit; color: var(--md-code-fg-color); font-feature-settings: &quot;kern&quot;; font-family: var(--md-code-font-family); direction: ltr; font-variant-ligatures: none; background-color: var(--md-code-bg-color); border-radius: 0.1rem; -webkit-box-decoration-break: clone; font-size: 0.85em; padding: 0px 0.294118em; word-break: break-word; -webkit-tap-highlight-color: transparent; outline: none;">{"&lt;option_name&gt;":"&lt;option_value&gt;"}</code>. Multiple options are separated with commas.</p><p style="box-sizing: inherit; color: rgb(0, 0, 0); font-family: Roboto, -apple-system, &quot;system-ui&quot;, Helvetica, Arial, sans-serif; font-size: 17.6px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Supported options of<span> </span><code style="box-sizing: inherit; color: var(--md-code-fg-color); font-feature-settings: &quot;kern&quot;; font-family: var(--md-code-font-family); direction: ltr; font-variant-ligatures: none; background-color: var(--md-code-bg-color); border-radius: 0.1rem; -webkit-box-decoration-break: clone; font-size: 0.85em; padding: 0px 0.294118em; word-break: break-word; -webkit-tap-highlight-color: transparent; outline: none;">rocksdb_db_options</code><span> </span>and<span> </span><code style="box-sizing: inherit; color: var(--md-code-fg-color); font-feature-settings: &quot;kern&quot;; font-family: var(--md-code-font-family); direction: ltr; font-variant-ligatures: none; background-color: var(--md-code-bg-color); border-radius: 0.1rem; -webkit-box-decoration-break: clone; font-size: 0.85em; padding: 0px 0.294118em; word-break: break-word; -webkit-tap-highlight-color: transparent; outline: none;">rocksdb_column_family_options</code><span> </span>are listed as follows.</p><ul style="box-sizing: inherit; margin-bottom: 1em; margin-top: 1em; list-style-type: disc; padding: 0px; margin-left: 0.625em; display: flow-root; color: rgb(0, 0, 0); font-family: Roboto, -apple-system, &quot;system-ui&quot;, Helvetica, Arial, sans-serif; font-size: 17.6px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: inherit; margin-bottom: 0px; margin-left: 1.25em;"><code style="box-sizing: inherit; color: var(--md-code-fg-color); font-feature-settings: &quot;kern&quot;; font-family: var(--md-code-font-family); direction: ltr; font-variant-ligatures: none; background-color: var(--md-code-bg-color); border-radius: 0.1rem; -webkit-box-decoration-break: clone; font-size: 0.85em; padding: 0px 0.294118em; word-break: break-word; -webkit-tap-highlight-color: transparent; outline: none;">rocksdb_db_options</code><div class="highlight" style="box-sizing: inherit;"><pre id="__code_1" style="box-sizing: inherit; color: var(--md-code-fg-color); font-feature-settings: &quot;kern&quot;; font-family: var(--md-code-font-family); margin-bottom: 1em; margin-top: 1em; direction: ltr; font-variant-ligatures: none; display: flow-root; line-height: 1.4; position: relative;"><span style="box-sizing: inherit;"></span><button class="md-clipboard md-icon" title="Copy to clipboard" data-clipboard-target="#__code_1 > code" style="box-sizing: inherit; -webkit-tap-highlight-color: transparent; background: transparent; border: 0px; font-family: inherit; font-size: inherit; margin: 0px; padding: 0px; border-radius: 0.1rem; color: var(--md-default-fg-color--lightest); cursor: pointer; height: 1.5em; outline: none; outline-offset: 0.1rem; position: absolute; right: 0.5em; top: 0.5em; transition: color 0.25s ease 0s; width: 1.5em; z-index: 1;"></button><code style="box-sizing: inherit; color: var(--md-code-fg-color); font-feature-settings: &quot;kern&quot;; font-family: var(--md-code-font-family); direction: ltr; font-variant-ligatures: none; background-color: var(--md-code-bg-color); border-radius: 0.1rem; -webkit-box-decoration-break: slice; font-size: 0.85em; padding: 0.772059em 1.17647em; word-break: normal; box-shadow: none; display: block; margin: 0px; outline: none; overflow: auto; touch-action: auto; -webkit-tap-highlight-color: transparent;">max_total_wal_size
delete_obsolete_files_period_micros
max_background_jobs
stats_dump_period_sec
compaction_readahead_size
writable_file_max_buffer_size
bytes_per_sync
wal_bytes_per_sync
delayed_write_rate
avoid_flush_during_shutdown
max_open_files
stats_persist_period_sec
stats_history_buffer_size
strict_bytes_per_sync
enable_rocksdb_prefix_filtering
enable_rocksdb_whole_key_filtering
rocksdb_filtering_prefix_length
num_compaction_threads
rate_limit
</code></pre></div></li></ul><ul style="box-sizing: inherit; margin-bottom: 1em; margin-top: 1em; list-style-type: disc; padding: 0px; margin-left: 0.625em; display: flow-root; color: rgb(0, 0, 0); font-family: Roboto, -apple-system, &quot;system-ui&quot;, Helvetica, Arial, sans-serif; font-size: 17.6px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: inherit; margin-bottom: 0px; margin-left: 1.25em;"><code style="box-sizing: inherit; color: var(--md-code-fg-color); font-feature-settings: &quot;kern&quot;; font-family: var(--md-code-font-family); direction: ltr; font-variant-ligatures: none; background-color: var(--md-code-bg-color); border-radius: 0.1rem; -webkit-box-decoration-break: clone; font-size: 0.85em; padding: 0px 0.294118em; word-break: break-word; -webkit-tap-highlight-color: transparent; outline: none;">rocksdb_column_family_options</code><div class="highlight" style="box-sizing: inherit;"><pre id="__code_2" style="box-sizing: inherit; color: var(--md-code-fg-color); font-feature-settings: &quot;kern&quot;; font-family: var(--md-code-font-family); margin-bottom: 1em; margin-top: 1em; direction: ltr; font-variant-ligatures: none; display: flow-root; line-height: 1.4; position: relative;"><span style="box-sizing: inherit;"></span><button class="md-clipboard md-icon" title="Copy to clipboard" data-clipboard-target="#__code_2 > code" style="box-sizing: inherit; -webkit-tap-highlight-color: transparent; background: transparent; border: 0px; font-family: inherit; font-size: inherit; margin: 0px; padding: 0px; border-radius: 0.1rem; color: var(--md-default-fg-color--lightest); cursor: pointer; height: 1.5em; outline: none; outline-offset: 0.1rem; position: absolute; right: 0.5em; top: 0.5em; transition: color 0.25s ease 0s; width: 1.5em; z-index: 1;"></button><code style="box-sizing: inherit; color: var(--md-code-fg-color); font-feature-settings: &quot;kern&quot;; font-family: var(--md-code-font-family); direction: ltr; font-variant-ligatures: none; background-color: var(--md-code-bg-color); border-radius: 0.1rem; -webkit-box-decoration-break: slice; font-size: 0.85em; padding: 0.772059em 1.17647em; word-break: normal; box-shadow: none; display: block; margin: 0px; outline: none; overflow: auto; touch-action: auto; -webkit-tap-highlight-color: transparent;">write_buffer_size
max_write_buffer_number
level0_file_num_compaction_trigger
level0_slowdown_writes_trigger
level0_stop_writes_trigger
target_file_size_base
target_file_size_multiplier
max_bytes_for_level_base
max_bytes_for_level_multiplier
disable_auto_compactions 
</code></pre></div></li></ul><br class="Apple-interchange-newline">


For more information, see [RocksDB official documentation](https://rocksdb.org/).

**Reply by wey-gu** [reply]
Also two URLs worthy to be shared:
- https://github.com/facebook/rocksdb/wiki/RocksDB-Tuning-Guide
- https://github.com/facebook/rocksdb/wiki/RocksDB-Troubleshooting-Guide

**Reply by wdroste** [reply]
For the default configurations is an template system size the team uses? Any specs along w/ the size of the database.

**Reply by wey-gu** [reply]
Thanks @wdroste , this is indeed missing.
@whitewum we could prepare more of this kind of reference in the Documentation/ Prod conf template?

**Reply by whitewum** [reply]
nothing can do with the configuration in this situation. Lookup just oom for a large dataset.

**Comment by wdroste** [comment]
Ok I up'd the specs quite a bit, 5 r6id.8xlarge

* 16 CPU
* 128 GB RAM
* 950 GB NVMe

Pushed 10m vertices with index on 5/10 properties. All the nodes are up and functioning and no crashes. The only thing is this LOOKUP query does not complete.

```
(root@nebula) [test]> LOOKUP ON Vulnerability YIELD id(vertex) | YIELD count(*);
[ERROR (-1005)]: Storage Error: part: 18, error: E_RPC_FAILURE(-3).

Thu, 27 Oct 2022 21:22:57 UTC
```

**Reply by wey-gu** [reply]
@critical27 correct me if wrong.

This RPC failure is not about RPC time out(where we could enlarge `storage_client_timeout_ms` to mitigate long queries triggered graphd --> storaged RPC time out), right? What could be the cause?

@wdroste would you mind sharing storaged logs at this time?
